Refactor LongPollingIT to have a narrower/stricter scope

npepinpe opened the pull request:
## Description

Refactors the LongPollingIT to be of narrower scope. The main idea is that the test now does the following:

- The first activate command is processed before the create process instance command
- Once the job created event is applied, a new job activate command should be sent again without the client having done so
- The client receives the job

To do that, we send the activate requests manually (to avoid the worker re-polling and thus getting the job without going through long polling). This ensures the client only sent a single request, and that if there are two ACTIVATE commands on the logs, it must have come from long polling.

Refactoring also includes:

- Simplification of set up (e.g. dumping logs of all nodes in a single extension, only using a single partition, etc.)
- Setting the long polling logger to trace for more information on failure
- Using `ContainerEngine` to be able to assert the ordering of the commands on the log, as well as getting the log printed out on failure

## Related issues

closes #9740 

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

oleschoenburg reviewed (commented):
Just pointing out that `shouldActivateAndCompleteJobsInTime` is a regression test for https://github.com/camunda/zeebe/issues/9658 and if I remember correctly, the low timeout was required to reproduce that bug.

Zelldon commented:
Totally agree with @oleschoenburg :100: so your statement ` we aren't testing performance here.` is not true :sweat_smile: but maybe we find a better way to test it :shrug:

npepinpe commented:
2 seconds is a very tight timeout for a QA test though, no? Is it really necessary for it to be so low?

npepinpe commented:
Is the worry that with a longer timeout the job worker will simply send a new request and we won't be going through the long polling path? In that case, why use a worker at all? Doesn't the normal `ActivateJobsRequest` use long polling anyway?

oleschoenburg commented:
> Is the worry that with a longer timeout the job worker will simply send a new request and we won't be going through the long polling path?

I think that was what we wanted to prevent with the low timeout, yes :+1:

npepinpe commented:
Do you remember why there are 3 partitions as well? Are we testing anything special there?

oleschoenburg commented:
I should have documented this better :grimacing: It does ring a bell but I don't remember it was ultimately required or not.

Zelldon commented:
I think it was less likely to happen with a single partition but not sure anymore :shrug:

npepinpe commented:
Wasn't the bug related to the advertised address? I don't really get why multiple partitions would have an effect there :thinking: 

So if I understand, the goal here is to test that when I send an activate jobs request _before_ creating a process instance, if new jobs come in, I will receive the jobs _before the activate jobs request times out_. And we want to guarantee that this is done via the notification/long polling, so we need to make sure:

1. The first activate command is processed _before_ the create process instance command
2. Once the job created event is applied, a new job activate command should be sent again _without the client having done so_
3. The client receives the job

Did I miss something?

npepinpe commented:
I've updated the test, and also the PR description about what I did.

Some downsides here is that the test is a little slower now, since we wait in two places for the engine to be "idle". I think it's also still a little too complicated, but it's also more accurate, so :shrug:

I added a `LoggersActuator` because I couldn't figure out how to configure specific loggers via environment variables without changing the log4j2 config, and that seems unsustainable (e.g. adding an env var for every possible logger :exploding_head:). If you know how though, I'd rather do this before starting the container than after via actuator.

oleschoenburg reviewed (approved):
When I revert https://github.com/camunda/zeebe/pull/9669/commits/dad58fedec29a71e3c4dcc8c6debd855137a035c and run the test it fails :+1: 

Code looks good to :+1:

npepinpe commented:
> Code looks good to :+1: 

I guess that's subjective, it's still a little too opaque/complex for my taste, but it'll do for now :smile:
